### ISSUE TITLE
mdc-textfield: make it possible to disable custom validity checking

### ISSUE THREAD

**Issue by gjdev** [material-components/material-components-web#1436]
MDCTextfieldFoundation offers a "custom validity checking" feature that makes it possible to handle validity checking by calling setValid(boolean) (pushing a new state) on the foundation, instead of via the adapter.getNativeInput().checkValidity() (the foundation pulls the validity state after certain events).

However, this is currently a one way street: It's possible to change from the "default validity checking" to "custom validity checking". But it's not possible (via the public API) to later change back to "default validity checking".

In frameworks like Vue.js, Angular, and React behavior is dynamically added to components, and can also be dynamically removed again. It therefore only makes sense to be able to change back from "custom validity checking" to "default validity checking".

We currently solve this in our Angular integration by calling (using the private API):

```
// disable custom validity checking:
foundation.useCustomValidityChecking_ = false;
// update the validity state of the textfield if affected by the new way to validate:
foundation.changeValidity_(adapter.getNativeInput().checkValidity());
```

I think the public API should offer a method to do just this.

**Comment by kfranqueiro** [comment]
The idea behind how this was implemented is that if you are using custom validity logic, the expectation is that you will be fully managing the validity of the text field yourself, and the validation logic of a text field is typically expected to remain consistent throughout its lifecycle.

Is your use case purely from a framework symmetry perspective? Do you foresee cases where this presents a concrete problem for end-users?

**Comment by gjdev** [comment]
It's a specific goal of this project to (quoted from the root readme) offer: "Seamless integration with other JS frameworks and libraries". In many JS frameworks properties of components are expected to be alterable continuously. I don't think it's very hard to come up with user based use cases, in which it makes sense to be able to do this as well. I.e. when an asynchronous validation returns "invalid", set the custom-validity to false. If after changes the asynchronous validation succeeds, start using the "normal" validity checking again.

It's literally two lines of code to make the component support this, and it (arguably I guess) makes for a better design.

(Too be honest I think the same goes for other properties on other components that are currently one-direction only. We are building an angular integration: https://blox.src.zone/material, and for a lot of components have to jump through hoops to make it possible to make all settings changeable in all directions. It's a lot easier to do this at the original component level, and imho makes for a better design and reusability of components, with no disadvantages I can think of).

**Comment by kfranqueiro** [comment]
I can at least understand the desire for this feature from the point of view of making framework integration as smooth as possible. Given that the changes sound minimal, we'd be willing to review a PR that adds this along with updated tests, if you'd like to create one.

Thanks!

**Comment by gjdev** [comment]
Thanks! I'll squeeze that in as work for our next sprint. So I expect to have a PR in around 2,5 weeks.

**Comment by gjdev** [comment]
It's delayed a bit, but stil in the pipeline.

**Comment by kfranqueiro** [comment]
I believe this has been resolved via #3084 back in July 2018 (present in 0.39 and higher). This is now controllable via the component's `useNativeValidation` setter.
